test/helper/common.bash

```diff
@@ -1,247 +1,397 @@
+    then
       echo "Timed out on command: ${*}" >&2
       return 1
     fi
   done
 }
 
-# @param ${1} timeout
-# @param ${2} container name
-# @param ... test command for container
-function repeat_in_container_until_success_or_timeout() {
-  local TIMEOUT="${1}"
-  local CONTAINER_NAME="${2}"
-  shift 2
+# ? << Functions about executing commands with timeouts
+# ! -------------------------------------------------------------------
+# ? >> Functions to wait until a condition is met
 
-  repeat_until_success_or_timeout --fatal-test "container_is_running ${CONTAINER_NAME}" "${TIMEOUT}" docker exec "${CONTAINER_NAME}" "${@}"
-}
 
-function container_is_running() {
-  [[ "$(docker inspect -f '{{.State.Running}}' "${1}")" == "true" ]]
-}
+# Wait until a port is ready.
+#
+# @param ${1} = port
+# @param ${2} = container name [OPTIONAL]
+function _wait_for_tcp_port_in_container() {
+  local PORT=${1:?Port number must be provided}
+  local CONTAINER_NAME=$(__handle_container_name "${2:-}")
 
-# @param ${1} port
-# @param ${2} container name
-function wait_for_tcp_port_in_container() {
-  repeat_until_success_or_timeout --fatal-test "container_is_running ${2}" "${TEST_TIMEOUT_IN_SECONDS}" docker exec "${2}" /bin/sh -c "nc -z 0.0.0.0 ${1}"
+  _repeat_until_success_or_timeout \
+    --fatal-test "_container_is_running ${CONTAINER_NAME}" \
+    "${TEST_TIMEOUT_IN_SECONDS}" \
+    _exec_in_container_bash "nc -z 0.0.0.0 ${PORT}"
 }
 
-# @param ${1} name of the postfix container
-function wait_for_smtp_port_in_container() {
-  wait_for_tcp_port_in_container 25 "${1}"
+# Wait for SMTP port (25) to become ready.
+#
+# @param ${1} = name of the container [OPTIONAL]
+function _wait_for_smtp_port_in_container() {
+  local CONTAINER_NAME=$(__handle_container_name "${1:-}")
+  _wait_for_tcp_port_in_container 25
 }
 
-# @param ${1} name of the postfix container
-function wait_for_smtp_port_in_container_to_respond() {
+# Wait until the SMPT port (25) can respond.
+#
+# @param ${1} = name of the container [OPTIONAL]
+function _wait_for_smtp_port_in_container_to_respond() {
+  local CONTAINER_NAME=$(__handle_container_name "${1:-}")
+
   local COUNT=0
-  until [[ $(docker exec "${1}" timeout 10 /bin/sh -c "echo QUIT | nc localhost 25") == *"221 2.0.0 Bye"* ]]; do
-    if [[ $COUNT -eq 20 ]]
+  until [[ $(_exec_in_container timeout 10 /bin/bash -c 'echo QUIT | nc localhost 25') == *'221 2.0.0 Bye'* ]]
+  do
+    if [[ ${COUNT} -eq 20 ]]
     then
       echo "Unable to receive a valid response from 'nc localhost 25' within 20 seconds"
       return 1
     fi
 
     sleep 1
-    ((COUNT+=1))
+    (( COUNT += 1 ))
   done
 }
 
-# @param ${1} name of the postfix container
-function wait_for_amavis_port_in_container() {
-  wait_for_tcp_port_in_container 10024 "${1}"
-}
+# Checks whether a service is running inside a container (${1}).
+#
+# @param ${1} = service name
+# @param ${2} = container name [OPTIONAL]
+function _container_has_service_running() {
```

Comment:
If a helper contains a `run` call, I think it should be more explicit on the function name with `_should` or `_run` prefix as a convention? It's ok to use `assert_` methods that don't depend on `run` context though.

```suggestion
function _should_have_service_running_in_container() {
```

Then again we have `_add_mail_account_then_wait_until_ready()` and that is more clearly an action and not something we'd use a `run` on when calling I think? So perhaps this is fine as-is :man_shrugging:

Reply:
I think the name you proposed is better; I will update it.